public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx

```diff
@@ -0,0 +1,356 @@
+
+import { stateHistoryApi } from '../../../api/stateHistoryApi';
+import { combineMatcherStrings, labelsMatchMatchers, parseMatchers } from '../../../utils/alertmanager';
+import { HoverCard } from '../../HoverCard';
+
+import { LogRecordViewerByTimestamp } from './LogRecordViewer';
+import { LogTimelineViewer } from './LogTimelineViewer';
+import { extractCommonLabels, Line, LogRecord, omitLabels } from './common';
+
+interface Props {
+  ruleUID: string;
+}
+
+const MAX_TIMELINE_SERIES = 20;
+
+const LokiStateHistory = ({ ruleUID }: Props) => {
+  const styles = useStyles2(getStyles);
+  const [instancesFilter, setInstancesFilter] = useState('');
+  const logsRef = useRef<HTMLDivElement[]>([]);
+
+  const { getValues, setValue, register, handleSubmit } = useForm({ defaultValues: { query: '' } });
+
+  const { useGetRuleHistoryQuery } = stateHistoryApi;
+  const timeRange = useMemo(() => getDefaultTimeRange(), []);
+  const {
+    currentData: stateHistory,
+    isLoading,
+    isError,
+    error,
+  } = useGetRuleHistoryQuery({ ruleUid: ruleUID, from: timeRange.from.unix(), to: timeRange.to.unix() });
+
+  const { dataFrames, historyRecords, commonLabels, totalRecordsCount } = useInstanceHistoryRecords(
+    stateHistory,
+    instancesFilter
+  );
+
+  const frameSubset = useMemo(() => take(dataFrames, MAX_TIMELINE_SERIES), [dataFrames]);
+
+  const onLogRecordLabelClick = useCallback(
+    (label: string) => {
+      const matcherString = combineMatcherStrings(getValues('query'), label);
+      setInstancesFilter(matcherString);
+      setValue('query', matcherString);
+    },
+    [setInstancesFilter, setValue, getValues]
+  );
+
+  const onFilterCleared = useCallback(() => {
+    setInstancesFilter('');
+    setValue('query', '');
+  }, [setInstancesFilter, setValue]);
+
+  const onTimelinePointerMove = useCallback(
+    (seriesIdx: number, pointIdx: number) => {
+      const uniqueTimestamps = sortBy(uniq(frameSubset.flatMap((frame) => frame.fields[0].values.toArray())));
+
+      const timestamp = uniqueTimestamps[pointIdx];
+
+      const refToScroll = logsRef.current.find((x) => parseInt(x.id, 10) === timestamp);
+      refToScroll?.scrollIntoView({ behavior: 'smooth', block: 'start' });
+    },
+    [frameSubset]
+  );
+
+  if (isLoading) {
+    return <div>Loading...</div>;
+  }
+  if (isError) {
+    return (
+      <Alert title="Error fetching the state history" severity="error">
+        {error instanceof Error ? error.message : 'Unable to fetch alert state history'}
+      </Alert>
+    );
+  }
+
+  const hasMoreInstances = frameSubset.length < dataFrames.length;
+  const emptyStateMessage =
+    totalRecordsCount > 0
+      ? `No matches were found for the given filters among the ${totalRecordsCount} instances`
+      : 'No state transitions have occurred in the last 60 minutes';
+
+  return (
+    <div className={styles.fullSize}>
+      <form onSubmit={handleSubmit((data) => setInstancesFilter(data.query))}>
```

Comment:
This would be a nice candidate for a separate component.

Reply:
Yep, there were a few issues, but I've been able to extract the Input component.